lib/multi_reduce_quda.cu

```diff
@@ -654,39 +636,43 @@ namespace quda {
         std::vector<ColorSpinorField*> w1(w.begin() + w.size()/2, w.end());
         multiReduce_recurse<doubleN,ReduceType,ReducerDiagonal,writeDiagonal,ReducerOffDiagonal,writeOffDiagonal>(result0, x, y0, z, w0, i_idx, 2*j_idx+0, hermitian, tile_size);
         multiReduce_recurse<doubleN,ReduceType,ReducerDiagonal,writeDiagonal,ReducerOffDiagonal,writeOffDiagonal>(result1, x, y1, z, w1, i_idx, 2*j_idx+1, hermitian, tile_size);
-      }
-      else
-      {
+      } else {
         doubleN* tmp_dot = new doubleN[x.size()*y.size()];
 
 	// if at bottom of recursion, return if on lower left
-	if (x.size() <= tile_size && hermitian) {
-	  if (j_idx < i_idx) { return; }
-	}
+        if (x.size() <= tile_size.x && is_valid_NXZ(x.size(), true) && hermitian) {
+          if (j_idx < i_idx) { return; }
+        }
 
         coeff_array<Complex> a, b, c;
 
-        if (x.size() <= tile_size) {
-        switch(x.size()){ // COMMENT HERE FOR COMPILE TIME
-        case 1:
-          multiReduce<1, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
+        if (x.size() <= tile_size.x && is_valid_NXZ(x.size(), true)) {
+          switch (x.size()) { // COMMENT HERE FOR COMPILE TIME
+          case 1:
+            multiReduce<1, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
               tmp_dot, a, b, c, x, y, z, w, i_idx, j_idx);
-          break;
-#if MAX_MULTI_BLAS_N >= 2
-        case 2:
-          multiReduce<2, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
+            break;
+          case 2:
+            multiReduce<2, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
               tmp_dot, a, b, c, x, y, z, w, i_idx, j_idx);
-          break;
+            break;
+          case 4:
+            multiReduce<4, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
+              tmp_dot, a, b, c, x, y, z, w, i_idx, j_idx);
+            break;
+          case 8:
+            multiReduce<8, doubleN, ReduceType, ReducerDiagonal, writeDiagonal, ReducerOffDiagonal, writeOffDiagonal, false>(
+              tmp_dot, a, b, c, x, y, z, w, i_idx, j_idx);
+            break;
+          case 16:
```

Comment:
Is there a reason we're only going up to 16 for reduce while we go up to 128 for regular blas? Compile time?

Also, is there a reason we always compile powers of two for blas independent of `MAX_MULTI_BLAS_N`, while we only compile powers of two for `MAX_MULTI_BLAS_N` > 2 in the reduction case?

Reply:
Disregard, my mind spaced out and I missed that the ` > 2` line was _removed_.

Reply:
> Is there a reason we're only going up to 16 for reduce while we go up to 128 for regular blas? Compile time?
A few reasons:
* Yes, partly compile time
* For 1-d multi-reductions, there is no need to have a high `NXZ` dimension since we always do this as a `1xn` reduction
* Also for 32, the compiler spills likes crazy and performance is bad so I figured only put it in if you really want it (which means let the user explicitly opt in for it, e.g., if they are using 2-d multi-reductions in a big block solver)